Some janitoring over the layout crate

nox opened the pull request:
This PR moves a bunch of code around and makes some methods from one-use traits into inherent methods.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/servo/servo/22684)
<!-- Reviewable:end -->

nox committed:
Make some cosmetic changes to ReflowGoal methods

nox committed:
Rename NewLayoutThreadInfo to LayoutThreadInit

Following the convention of our other init structs.

nox committed:
Move CSSReporter from script_layout_interface to script

nox committed:
Remove BlockFlowDisplayListBuilding

Replaced by inherent methods on BlockFlow, the only implementor of that trait.

nox committed:
Remove BaseFlowDisplayListBuilding

nox committed:
Remove FragmentDisplayListBuilding

nox committed:
Kill establishes_containing_block_for_absolute

We inline the related tests at the two call sites.

nox committed:
Remove InlineFlowDisplayListBuilding

This lets us remove a few call indirections where a function just calls
another function defined in another module.

nox committed:
Remove ListItemFlowDisplayListBuilding

nox committed:
Kill RgbColor

nox committed:
Remove FlexFlowDisplayListBuilding

highfive commented:
Heads up! This PR modifies the following files:
 * @asajeffrey: components/script/dom/bindings/trace.rs, components/script/script_thread.rs, components/script/dom/window.rs
 * @kichjang: components/script/dom/bindings/trace.rs, components/script/script_thread.rs, components/script/dom/window.rs
 * @emilio: components/layout/table_caption.rs, components/layout/table.rs, components/layout/display_list/builder.rs, components/layout/table_wrapper.rs, components/layout/table_rowgroup.rs and 7 more

highfive commented:
<img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20> **Warning** <img src="http://www.joshmatthews.net/warning.svg" alt="warning" height=20>

* These commits modify layout and script code, but no tests are modified. Please consider adding a test!

nox commented:
r? @SimonSapin

nox committed:
Remove FlowConstructionUtils

nox committed:
Remove LayoutDamageComputation

nox committed:
Remove CompletelyEncloses

nox committed:
Remove ComputedValuesCursorUtility

nox committed:
Remove ComputedValueUtils

nox committed:
Remove ObjectElement

nox committed:
Make Window::scroll_offsets store keys as OpaqueNode values

This is the type that is supposed to signal that we will never ever
try to get back a Node from it in an unsafe way, unlike
UntrustedNodeAddress.

emilio commented on components/script/dom/window.rs:
```diff
@@ -2242,3 +2241,41 @@ impl Window {
+            .unwrap()
+            .send(ConstellationControlMsg::ReportCSSError(
+                self.pipelineid,
+                "".to_owned(),
```
We have the URL here so reporting it should be trivial, mind filing an issue for that?

nox replied:
I drive-by fixed it.

emilio commented on components/layout/display_list/builder.rs:
```diff
@@ -2360,11 +2347,11 @@ impl BlockFlow {
-            self.positioning(),
-            established_reference_frame.is_some(),
-        ) {
+        if established_reference_frame.is_some() ||
```
can you do instead:

```rust
let abspos_containing_block = established_reference_frame.is_some() || ....;
if abspos_containing_block {
```

emilio commented on components/layout/display_list/builder.rs:
```diff
@@ -2842,11 +2829,7 @@ impl InlineFlowDisplayListBuilding for InlineFlow {
-                fragment.style.get_box().position,
-                false,
-            ) {
+            if fragment.style.get_box().position != StylePosition::Static {
```
And same here?

This looks wrong btw, an inline can also establish a reference frame.

nox replied:
Want me to file a follow-up issue? This is just the same code as before right?

emilio commented on components/layout/incremental.rs:
```diff
@@ -22,13 +22,8 @@ bitflags! {
-
-impl<'a> LayoutDamageComputation for &'a mut dyn Flow {
-    fn compute_layout_damage(self) -> SpecialRestyleDamage {
+impl dyn Flow {
```
Oh, this is fancy.

emilio commented on components/layout/display_list/builder.rs:
```diff
@@ -2828,30 +2826,24 @@ impl BaseFlow {
+/// text display items it may be `TextCursor` or `VerticalTextCursor`.
+#[inline]
+fn get_cursor(values: &ComputedValues, default_cursor: CursorKind) -> Option<CursorKind> {
+    match (
```
You could also just move it to style, your call.

In any case I think this is simpler to read as follows, if you want to make that change:

```rust
let ui = values.get_inherited_ui();
if ui.pointer_events == PointerEvents::None {
    return None;
}

match ui.cursor {
    // ...
}
```

nox replied:
Made some changes like you suggested but didn't move it to style given there is no other user.

nox committed:
Report the stylesheet URL when reporting CSS errors

nox committed:
Simplify get_cursor

CYBAI commented:
From Travis 👀

```
Diff in /home/travis/build/servo/servo/components/script/dom/window.rs at line 1602:
     }
 
     pub fn scroll_offset_query(&self, node: &Node) -> Vector2D<f32> {
-        if let Some(scroll_offset) = self
-            .scroll_offsets
-            .borrow()
-            .get(&node.to_opaque())
-        {
+        if let Some(scroll_offset) = self.scroll_offsets.borrow().get(&node.to_opaque()) {
             return *scroll_offset;
         }
         Vector2D::new(0.0, 0.0)
Diff in /home/travis/build/servo/servo/components/script/dom/window.rs at line 1621:
         // The scroll offsets are immediatly updated since later calls
         // to topScroll and others may access the properties before
         // webrender has a chance to update the offsets.
-        self.scroll_offsets.borrow_mut().insert(
-            node.to_opaque(),
-            Vector2D::new(x_ as f32, y_ as f32),
-        );
+        self.scroll_offsets
+            .borrow_mut()
+            .insert(node.to_opaque(), Vector2D::new(x_ as f32, y_ as f32));
 
         let NodeScrollIdResponse(scroll_id) = self.layout_rpc.node_scroll_id();
 
Run `./mach fmt` to fix the formatting
```

nox commented:
@bors-servo r=emilio,SimonSapin

bors-servo commented:
:pushpin: Commit cf15336 has been approved by `emilio,SimonSapin`

<!-- @bors-servo r=emilio,SimonSapin cf15336a51af938b1cc6d9f161d0ebdd6abda36b -->

bors-servo commented:
:hourglass: Testing commit cf15336a51af938b1cc6d9f161d0ebdd6abda36b with merge c242abd99931676f246152daadf2b8dd9ddf8177...

bors-servo commented:
:sunny: Test successful - [android-mac](http://build.servo.org/builders/android-mac/builds/957), [arm32](http://build.servo.org/builders/arm32/builds/13829), [arm64](http://build.servo.org/builders/arm64/builds/13857), [linux-rel-css](http://build.servo.org/builders/linux-rel-css/builds/11455), [linux-rel-wpt](http://build.servo.org/builders/linux-rel-wpt/builds/11511), [mac-rel-css1](http://build.servo.org/builders/mac-rel-css1/builds/6435), [mac-rel-css2](http://build.servo.org/builders/mac-rel-css2/builds/6401), [mac-rel-wpt1](http://build.servo.org/builders/mac-rel-wpt1/builds/10651), [mac-rel-wpt2](http://build.servo.org/builders/mac-rel-wpt2/builds/10709), [mac-rel-wpt3](http://build.servo.org/builders/mac-rel-wpt3/builds/6461), [mac-rel-wpt4](http://build.servo.org/builders/mac-rel-wpt4/builds/6389), [magicleap](http://build.servo.org/builders/magicleap/builds/576), [status-taskcluster](https://tools.taskcluster.net/task-group-inspector/#/Dl73v0H6QZ-YXEFjzoN9Aw)
Approved by: emilio,SimonSapin
Pushing c242abd99931676f246152daadf2b8dd9ddf8177 to master...
